src/state_machine.zig

```diff
@@ -305,10 +305,10 @@ pub const StateMachine = struct {
         assert(t.timestamp > self.commit_timestamp);
 
         if (t.flags.padding != 0) return .reserved_flag_padding;
-        if (t.flags.two_phase_commit) {
+        if (t.flags.two_phase_commit and t.timeout == 0) {
             // Otherwise reserved amounts may never be released:
-            if (t.timeout == 0) return .two_phase_commit_must_timeout;
-        } else if (t.timeout != 0) {
+            return .two_phase_commit_must_timeout;
+        } else if (!t.flags.two_phase_commit and t.timeout != 0) {
```

Comment:
Let's branch on the major two-phase decision first as we had it rather. This keeps the validations for the two types of transfer separate and also ensures that we don't check the `two_phase_commit` flag again, even if the compiler will probably optimize that (at least we don't depend on that).